Fix errors not matching PurchasesError

vegaro opened the pull request:
We have this long standing issue #302 that reported that our errors don't match the typescript interface, particularly in iOS.

The main reason is that in Android, Capacitor automatically puts whatever is passed as the third parameter of the reject function in a `data` property in the error. In iOS, we were not even passing the user info, so the errors looked like this:

### Before the change

in Android:

```
{
  "code": "11",
  "data": {
    "code": 11,
    "message": "There was a credentials issue. Check the underlying error for more details.",
    "readableErrorCode": "InvalidCredentialsError",
    "readable_error_code": "InvalidCredentialsError",
    "underlyingErrorMessage": "Invalid API Key."
  }
}
```

in iOS:
```
{
  "code":"11",
  "errorMessage":"There was a credentials issue. Check the underlying error for more details. Invalid API Key.",
  "message":"There was a credentials issue. Check the underlying error for more details. Invalid API Key."
}
```

### After the change

in Android:

```
{
  "code": "11",
  "data": {
    "code": 11,
    "message": "There was a credentials issue. Check the underlying error for more details.",
    "readableErrorCode": "InvalidCredentialsError",
    "readable_error_code": "InvalidCredentialsError",
    "underlyingErrorMessage": "Invalid API Key."
  },
  "message": "There was a credentials issue. Check the underlying error for more details.",
  "readableErrorCode": "InvalidCredentialsError",
  "userInfo": {
    "readableErrorCode": "InvalidCredentialsError"
  },
  "underlyingErrorMessage": "Invalid API Key.",
  "userCancelled": null
}
```

in iOS:
```
{
  "code": "11",
  "data": {
    "message": "There was a credentials issue. Check the underlying error for more details. Invalid API Key.",
    "readable_error_code": "INVALID_CREDENTIALS",
    "code": 11,
    "readableErrorCode": "INVALID_CREDENTIALS",
    "underlyingErrorMessage": "Invalid API Key."
  },
  "errorMessage": "There was a credentials issue. Check the underlying error for more details. Invalid API Key.",
  "message": "There was a credentials issue. Check the underlying error for more details. Invalid API Key.",
  "readableErrorCode": "INVALID_CREDENTIALS",
  "userInfo": {
    "readableErrorCode": "INVALID_CREDENTIALS"
  },
  "underlyingErrorMessage": "Invalid API Key.",
  "userCancelled": null
}
```

### Related PRs

One change across three repos:

- RevenueCat/purchases-hybrid-common#1838: `normalizePurchasesError` and `withNormalizedErrors`, plus the iOS side
- RevenueCat/purchases-capacitor#579: Capacitor wiring, taken over from @vegaro
- RevenueCat/react-native-purchases#1919: React Native wiring

These merge together. Releasing purchases-hybrid-common on its own would let the automatic bump into react-native-purchases land ahead of the fix there, dropping `userInfo.readableErrorCode` for React Native iOS consumers.

Tracked in SDK-4450.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> **Medium Risk**
> Changes the JS error object apps catch from native calls, which can break existing error handling. Also pins PHC to a local tarball that must not ship as-is.
> 
> **Overview**
> Native plugin errors now match the `PurchasesError` TypeScript shape instead of Capacitor’s raw reject payload (especially on iOS).
> 
> iOS `rejectWithErrorContainer` now passes `error.info` as the reject data so JS gets `readableErrorCode`, `underlyingErrorMessage`, and related fields. Both `Purchases` and `RevenueCatUI` wrap `registerPlugin` with `withNormalizedErrors` so those fields are copied onto the thrown `Error` while keeping the original `data` payload.
> 
> Adds Jest coverage for that normalization. Internal PHC is temporarily pointed at a local `18.33.0` tarball.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 916bdec9832b915c9b5c91338b0438fb30e59549. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

ajpallares reviewed (commented):
I think it makes sense! 

Should we add some unit tests for this?

Also, I'll leave approval to someone with more ts knowledge 🙏

tonidero reviewed (approved):
Just a question, but if we've tested it with most functions on the SDK in both Android and iOS, it does make sense. Thank you so much for handling this!!

vegaro commented:
I should check this still works fine with Capacitor 8 before merging

vegaro committed:
fix(ios): Send the error info payload to the JS layer

Android already passes ErrorContainer.info as the reject data; iOS dropped it,
so readableErrorCode and underlyingErrorMessage never reached consumers.

Co-authored-by: Álvaro Brey <alvaro.brey@revenuecat.com>

AlvaroBrey committed:
chore: Point at a local purchases-hybrid-common build

Temporary. Replace with the published version once the normalizer ships in
purchases-hybrid-common; drop the local-phc directory with it.

AlvaroBrey committed:
fix: Make errors match the PurchasesError interface

The bridge nests the error payload under data, so readableErrorCode,
underlyingErrorMessage, userCancelled and userInfo never appeared where the
interface declares them.

Co-authored-by: Cesar de la Vega <664544+vegaro@users.noreply.github.com>

AlvaroBrey committed:
test: Cover error normalization at the plugin boundary

AlvaroBrey commented on package.json:
```diff
@@ -89,6 +93,6 @@
   },
   "dependencies": {
-    "@revenuecat/purchases-typescript-internal-esm": "18.32.1"
+    "@revenuecat/purchases-typescript-internal-esm": "file:./local-phc/revenuecat-purchases-typescript-internal-esm-18.33.0-local.0.tgz"
```
This is only for the draft, this PR will not be merged before the appropriate PHC bump

AlvaroBrey committed:
chore: Refresh the local purchases-hybrid-common build

AlvaroBrey committed:
refactor: Stop wrapping the RevenueCatUI plugin

Its rejections carry names such as PAYWALL_ERROR rather than error codes, so
normalization is a no-op there.

AlvaroBrey committed:
refactor: Hoist the paywall impression wrapper and use the real CapacitorException

AlvaroBrey committed:
ci: Run the unit tests

AlvaroBrey committed:
chore: Refresh the local purchases-hybrid-common build

Drops the unnecessary yarn build from the unit test job; ts-jest compiles from
source and nothing reads dist.

AlvaroBrey committed:
test: Pin the exact error shape consumers see

AlvaroBrey committed:
chore: Sync the UI package lockfile with the local build

AlvaroBrey committed:
chore: Refresh the local purchases-hybrid-common build
